### ISSUE TITLE
Standardizing file extensions

### ISSUE THREAD

**Issue by jonathanadler** [css-modules/css-modules#229]
I would like to make a proposal and I would like to know what you all think.
I wasn't 100% percent this is the place to suggest this, but I thought it might be a good place to start.

CSS Modules are awesome. My team and I use it in all our projects.
But the lack of standardization, makes it hard, at times, especially when working with external packages.

So we use the extension: _*.icss_ to indicate that the file is used as a module.

However, we mostly use less as our preprocessor, but also import packages that use sass (or scss).
This causes some problems with the way we load files.

So our suggestion is to create a standard for style files' extensions, which is:
- files with the regular extensions (_*.css_, _*.less_, _*.sass_, _*.scss_, etc.) indicate a regular cascading style file which is not scoped.
- In order to indicate that a file is used as a module, just add an _i_ before the extension (_*.icss_, _*.iless_, _*.isass_, _*.iscss_, etc.) - it could be something else of course...

IMHO, If this was a standard, that babel and webpack loaders could align with, our lives could become a bit easier :)

I'd be glad to hear what you think.

Thanks!

**Comment by kud** [comment]
I'm really interested by it but why "i"? I'm curious.

**Comment by jonathanadler** [comment]
@kud 
I believe it came from this:
[icss](https://github.com/css-modules/icss)

**Comment by kud** [comment]
It's interesting but I think calling them xxx.local.css is more efficient as it doesn't change the format and so your editor. I've got icon and syntax on css by default where I don't with `.icss`.

**Comment by jonathanadler** [comment]
That's true.
However, with `.local.css`, loaders will read them as regular css files, which is the main reason I'm proposing this change.
That being said, I'm not at all attached to the _i_ prefix solution.

**Comment by kud** [comment]
Maybe I'm wrong but isn't it just a question of regex here, no?

For the moment I do:

```
        // styles
        {
          test: /\.css$/,
          exclude: [path.resolve(__dirname, 'src', 'components'), path.resolve(__dirname, 'src', 'pages')],
          use: [
            {
              loader: 'style-loader'
            },
            {
              loader: 'css-loader',
              options: {
                modules: false,
                importLoaders: 1,
                import: true
              }
            },
            postCssLoader
          ]
        },

        {
          test: /\.css$/,
          include: [path.resolve(__dirname, 'src', 'components'), path.resolve(__dirname, 'src', 'pages')],
          use: [
            {
              loader: 'style-loader'
            },
            {
              loader: 'css-loader',
              options: {
                modules: true,
                importLoaders: 1,
                localIdentName: env === 'prod' ? '[name]--[local]___[hash:base64:5]' : '[path][name]--[local]___[hash:base64:5]',
                import: true
              }
            },
            postCssLoader
          ]
        },
```

I could remove `include` and `exclude` and use `.local.css` and `.css` instead (where I don't match .local.css in the second part).

**Comment by kud** [comment]
In any case, we need two matches, right?

**Comment by jonathanadler** [comment]
That's totally correct.
For me, both suggestions work fine, as long as it's a standard.

**Comment by kud** [comment]
Yes :) thank you. I'm trying right now with `.local.css`. I'll see if it works well for me. Thank you for your suggestion, it's a good idea.

**Comment by kud** [comment]
My current solution:


```javascript
const postCssLoader = {
    loader: 'postcss-loader',
    options: {
      plugins: [
        // autoprefixer(), unneeded for the moment
        customMedia({
          extensions: {
            "--mobile": "(max-width: 480px)",
            "--mobile-only": "(max-width: 480px)",
            "--tablet": "(max-width: 1024px)",
            "--tablet-only": "(min-width: 481px) and (max-width: 1024px)",
            "--desktop": "(min-width: 1025px)",
            "--desktop-only": "(min-width: 1025px)",
          }
        })
      ]
    }
  }

        // styles
        {
          test: /\.css$/,
          exclude: /\.local\.css$/,
          use: [
            {
              loader: 'style-loader'
            },
            {
              loader: 'css-loader',
              options: {
                modules: false,
                importLoaders: 1,
                import: true
              }
            },
            postCssLoader
          ]
        },

        {
          test: /\.local\.css$/,
          use: [
            {
              loader: 'style-loader'
            },
            {
              loader: 'css-loader',
              options: {
                modules: true,
                importLoaders: 1,
                localIdentName: env === 'prod' ? '[name]--[local]___[hash:base64:5]' : '[path][name]--[local]___[hash:base64:5]',
                import: true
              }
            },
            postCssLoader
          ]
        },
```

I'll refactor asap I think.

**Comment by taion** [comment]
Wouldn't something like `.mcss` by analogy with `.mjs` be the most consistent?

**Comment by TrySound** [comment]
@taion mjs wan't introduced. We are left with `.js`. BTW maybe it's a good idea for css too? css modules are not specfial syntax. It's regular css with a few benefits.

**Comment by kud** [comment]
`.local.css` is quite clear, and it's the term used in css modules like

```css
:local(.myClass)
```

**Comment by borela** [comment]
`.local.css` also makes sense because it's the default visibility for CSS modules.

**Comment by taion** [comment]
I thought `.mjs` was still the plan for Node?

**Comment by TrySound** [comment]
@taion Don't think so. Modules should be a regular js. cjs or anything is a legacy.

**Comment by taion** [comment]
Please double-check. I believe the current plan is ES module code will still be `.mjs`.

**Comment by TrySound** [comment]
It's a fallback, the last solution, not main.

**Comment by ro-savage** [comment]
Create React App is looking at using `[name].module.css` in [PR2285](facebookincubator/create-react-app#2285)

This is because its explicit and does not require any changes to tooling (IDEs, linters, highlighters, etc)

**Comment by taion** [comment]
Perhaps still `.cssm` for consistency with the upcoming `cssm-loader` for Webpack, then?

**Comment by ro-savage** [comment]
@taion - what is `cssm-loader` I can't find anything about it?

**Comment by kud** [comment]
I'm using `[name].local.css` and it works like a charm. It means what it means. But `module` could be ok too.

**Comment by jonathanadler** [comment]
After going over the thread, you guys convinced me that adding ```.local.css``` or ```.module.css``` would be the best way to go, and not changing the file extension.
Regarding both suggestions, I find that @ro-savage's suggestion ```.module.css``` will be more understood since it's quite straightforward.  ```local``` is a bit more ambiguous.

**Comment by borela** [comment]
To me `.local` is more intuitive as it expresses what is happening to the CSS classes in that particular file, `.module` would make more sense for bundles/libs.

**Comment by ro-savage** [comment]
I think you are looking at it only from the point of view of people who are already using CSS Modules.

Yes it is locally scoped, but there are other methods to locally scope CSS both now and in the future.  `.local.css` may give a hint its scoped but it doesn't help understand what the heck this file is.

`.module.css` gives a hint this is a CSS Module which they can then research to find out that its locally scoped, how it works, etc.

People who use CSS modules and are creating `.module.css` or `.local.css` files already know how it works. What you want to worry about are the people who **don't** know what it is or how it works.

Its a bit of a flimsy argument. But the point is that the file extension should related to the tool that understands it (or framework in this case) rather than what it does, because lots of things locally scope CSS but only CSS Modules is CSS Modules.

**Comment by borela** [comment]
@ro-savage 
> there are other methods to locally scope CSS both now and in the future. .local.css may give a hint its scoped but it doesn't help understand what the heck this file is.

That made sense, I forgot to consider the other methods.

**Comment by kud** [comment]
Plus, if you read right to left, it does "css.modules." :D

**Comment by kud** [comment]
For those who are interested: 

```
$ rename s/local\.css/module\.css/ **/*.css
```

:)

I did it just right now.

**Comment by borela** [comment]
[I converted my webpack configs to use `.module.[css|sass|...]` too.](https://github.com/ctrine/webpack-settings/blob/master/packages/shared/src/index.js)

**Comment by kud** [comment]
Hey @markdalgleish, what do you think by standardising css modules filenames with `[name].module.css`?

**Comment by ahfarmer** [comment]
FYI [Gatsby](https://github.com/gatsbyjs/gatsby) uses `[name].module.css` in its default configuration. Gatsby is awesome and quite popular.

**Comment by petermikitsh** [comment]
Reviewing this discussion, it seems as if the community has settled on the extension `.module.css`. Is there anything else that could be done to make this _officially_ official? Perhaps a PR is in order, mentioning that the preferred file extension is `.module.css` in the README.
